examples/android/routeguide/app/src/main/java/io/grpc/routeguideexample/RouteGuideActivity.java

```diff
@@ -0,0 +1,402 @@
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class RouteGuideActivity extends ActionBarActivity {
+    private static final Logger logger = Logger.getLogger(RouteGuideActivity.class.getName());
+    private EditText mHostEdit;
+    private EditText mPortEdit;
+    private Button mGetFeatureButton;
+    private Button mListFeaturesButton;
+    private Button mRecordRouteButton;
+    private Button mRouteChatButton;
+    private TextView mResultText;
+
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_routeguide);
+        mHostEdit = (EditText) findViewById(R.id.host_edit_text);
+        mPortEdit = (EditText) findViewById(R.id.port_edit_text);
+        mGetFeatureButton = (Button) findViewById(R.id.get_feature_button);
+        mListFeaturesButton = (Button) findViewById(R.id.list_features_button);
+        mRecordRouteButton = (Button) findViewById(R.id.record_route_button);
+        mRouteChatButton = (Button) findViewById(R.id.route_chat_button);
+        mResultText = (TextView) findViewById(R.id.result_text);
+        mResultText.setMovementMethod(new ScrollingMovementMethod());
+    }
+
+    public void getFeature(View view) {
+        new GrpcTask(new GetFeatureRunnable()).execute();
+    }
+
+    public void listFeatures(View view) {
+        new GrpcTask(new ListFeaturesRunnable()).execute();
+    }
+
+    public void recordRoute(View view) {
+        new GrpcTask(new RecordRouteRunnable()).execute();
+    }
+
+    public void routeChat(View view) {
+        new GrpcTask(new RouteChatRunnable()).execute();
+    }
+
+    private class GrpcTask extends AsyncTask<Void, Void, String> {
+        private String mHost;
+        private String mMessage;
+        private int mPort;
+        private final GrpcRunnable mGrpc;
+        private ManagedChannel mChannel;
+
+        GrpcTask(GrpcRunnable grpc) {
+            this.mGrpc = grpc;
+        }
+
+        @Override
+        protected void onPreExecute() {
+            mHost = mHostEdit.getText().toString();
+            String portStr = mPortEdit.getText().toString();
+            mPort = TextUtils.isEmpty(portStr) ? 0 : Integer.valueOf(portStr);
+            mResultText.setText("");
+            ((InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE))
+                    .hideSoftInputFromWindow(mHostEdit.getWindowToken(), 0);
+            mGetFeatureButton.setEnabled(false);
+            mListFeaturesButton.setEnabled(false);
+            mRecordRouteButton.setEnabled(false);
+            mRouteChatButton.setEnabled(false);
+        }
+
+        @Override
+        protected String doInBackground(Void... nothing) {
+            try {
+                mChannel = ManagedChannelBuilder.forAddress(mHost, mPort)
```

Comment:
Then either we can use a different UI (have one Activity to choose, and one to try things), or we can notice that the host/port is different, shut down the old one, and create a new one.

The example needs to adhere to best practices, otherwise it can do more harm than good.